fedimint-client/src/sm/executor.rs

```diff
@@ -390,181 +405,258 @@ impl ExecutorInner {
+                                                dbtx.insert_entry(&k, &v).await;
+                                                dbtx.on_commit(move || {
+                                                    let _ = sm_update_tx.send(state);
+                                                });
+                                                Ok(ActiveOrInactiveState::Active {
+                                                    dyn_state: new_state,
+                                                    meta: v,
+                                                })
+                                            }
+                                        })
+                                    },
+                                    None,
                                 )
-                                .await;
-
-                                let context = &self
-                                    .module_contexts
-                                    .get(&state.module_instance_id())
-                                    .expect("Unknown module");
-
-                                let global_context = global_context_gen(
-                                    state.module_instance_id(),
-                                    state.operation_id(),
-                                );
-                                if new_state.is_terminal(context, &global_context) {
-                                    // TODO: log state machine id or something
-                                    debug!("State machine reached terminal state");
-                                    let k = InactiveStateKey::from_state(new_state.clone());
-                                    let v = ActiveState::new().into_inactive();
-                                    dbtx.insert_entry(&k, &v).await;
-                                    Ok(ActiveOrInactiveState::Inactive {
-                                        dyn_state: new_state,
-                                    })
-                                } else {
-                                    let k = ActiveStateKey::from_state(new_state.clone());
-                                    let v = ActiveState::new();
-                                    dbtx.insert_entry(&k, &v).await;
-                                    Ok(ActiveOrInactiveState::Active {
-                                        dyn_state: new_state,
-                                        active_state: v,
-                                    })
+                                .await
+                                .map_err(|e| match e {
+                                    AutocommitError::CommitFailed {
+                                        last_error,
+                                        attempts,
+                                    } => last_error.context(format!(
+                                        "Failed to commit after {attempts} attempts"
+                                    )),
```

Comment:
You use `None` as the number of retries, that means it will run forever, so this could be an `unreachable!`?

Reply:
I think so. BTW. I see `Some(100)` etc. all over the code, and I wonder if we have preferences.

Reply:
`None` is more correct, but potentially harder to notice issues with. I often chose `Some(100)` as a canary that will tell us if something is being retried excessively.